symbolic-base/src/ZkFold/Base/Protocol/Protostar/SpecialSound.hs

```diff
@@ -28,58 +28,36 @@ challenge ri ∈ F. After the final message mk, the verifier computes the algebr
-        -> VerifierMessage f a -- ^ current random challenge
-        -> Natural             -- ^ round number (starting from 0)
-        -> ProverMessage f a
-
-      verifier :: a
-        -> Input f a             -- ^ public input
-        -> [ProverMessage f a]   -- ^ prover messages
-        -> [VerifierMessage f a] -- ^ random challenges
-        -> VerifierOutput f a    -- ^ verifier output
-
-type BasicSpecialSoundProtocol f pi m a =
-  ( SpecialSoundProtocol f a
-  , Witness f a ~ ()
-  , Input f a ~ pi
-  , ProverMessage f a ~ m
-  , VerifierMessage f a ~ f
-  )
-
-instance (Arithmetic a, KnownNat n) => SpecialSoundProtocol a (ArithmeticCircuit a U1 (Vector n) o) where
-    type Witness a (ArithmeticCircuit a U1 (Vector n) o) = ()
-    type Input a (ArithmeticCircuit a U1 (Vector n) o) = [a]
-    type ProverMessage a (ArithmeticCircuit a U1 (Vector n) o) = [a]
-    type VerifierMessage a (ArithmeticCircuit a U1 (Vector n) o) = a
-    type VerifierOutput a (ArithmeticCircuit a U1 (Vector n) o)  = [a]
-    -- type Degree (ArithmeticCircuit a p (Vector n) o) = AM.Degree (ArithmeticCircuit a p (Vector n) o)
-
-    -- One round for Plonk
-    rounds = P.const 1
-
-    outputLength ac = P.fromIntegral $ M.size (acSystem ac)
-
-    -- The transcript will be empty at this point, it is a one-round protocol.
-    -- Input is arithmetised. We need to combine its witness with the circuit's witness.
-    --
-    prover ac _ pi _ _ = elems $ witnessGenerator ac U1 $ unsafeToVector pi
-
-    -- | Evaluate the algebraic map on public inputs and prover messages and compare it to a list of zeros
-    --
-    verifier ac pi pm ts = AM.algebraicMap ac pi pm ts one
+class SpecialSoundProtocol f i p m c d k a where
+  type VerifierOutput f i p m c d k a
+
+  input :: a
+    -> i f                          -- ^ previous public input
+    -> p f                          -- ^ witness
+    -> i f                          -- ^ public input
+
+  prover :: a
+    -> i f                          -- ^ previous public input
+    -> p f                          -- ^ witness
+    -> f                            -- ^ current random challenge
+    -> Natural                      -- ^ round number (starting from 1)
+    -> m                            -- ^ prover message
+
+  verifier :: a
+    -> i f                          -- ^ public input
+    -> Vector k m                   -- ^ prover messages
+    -> Vector (k-1) f               -- ^ random challenges
+    -> VerifierOutput f i p m c d k a -- ^ verifier output
+
+instance (Arithmetic a, Representable i, Representable p, KnownNat (d + 1))
+    => SpecialSoundProtocol a i p [a] c d 1 (ArithmetizableFunction a i p) where
+  type VerifierOutput a i p [a] c d 1 (ArithmetizableFunction a i p) = [a]
+
+  input ArithmetizableFunction {..} = afEval
+
+  -- | Just return the witness values on the previous public input
+  prover ArithmetizableFunction {..} pi0 w _ _ = elems $ witnessGenerator afCircuit (pi0 :*: w) undefined
```

Comment:
Instead of `undefined`, `afEval pi0 w` can be safely used here in case optimized circuit ends up using the guessed output in witness computation